natlib/svcd.c

```diff
@@ -185,3 +186,120 @@ static int svcd_init( lua_State *L )
 
     return 0;
 }
+
+//////////////////////////////////////////////////////////////////////////////
+// SVCD.notify() implementation
+// Maintainers: Sam Kumar <samkumar@berkeley.edu>
+//              Leonard Truong <leonardtruong@berkeley.edu>
+//              Michael Chen <mc.michaelchen.us@gmail.com>
+/////////////////////////////////////////////////////////////
+int notify_cord(lua_State* L) {
+    int header_index;
+    // Push SVCD.subscribers[svc_id][attr_id] (upvalue 1) onto stack
+    lua_pushvalue(L, lua_upvalueindex(1));
+
+    // iterate by key value pairs using lua_next
+    // see: http://www.lua.org/manual/5.1/manual.html#lua_next
+    for (lua_pushnil(L); lua_next(L, 1); lua_pop(L, lua_gettop(L) - 2)) {
+
+        // local header = storm.array.create(1, storm.array.UINT16)
+	lua_pushlightfunction(L, arr_create);
+	lua_pushnumber(L, 1);
+	lua_pushnumber(L, ARR_TYPE_UINT16);
+	lua_call(L, 2, 1);
+	header_index = lua_gettop(L);
+
+        // header:set(1, v)
+	lua_pushstring(L, "set");
+	lua_gettable(L, -2);
+	lua_pushnumber(L, 1);
+	lua_pushvalue(L, 3);  // Store current value (3rd element on stack)
+	lua_call(L, 2, 0);
+
+        // storm.net.sendto(SVCD.ncsock, header:as_str()..value, k, 2527)
+	lua_pushlightfunction(L, libstorm_net_sendto);
+	lua_getglobal(L, "SVCD");
+	lua_pushstring(L, "ncsock");
+	lua_gettable(L, -2);
+	lua_pushstring(L, "as_str");
+	lua_gettable(L, header_index);
+	lua_pushvalue(L, lua_upvalueindex(2));  // value stored as upvalue 2
+	lua_concat(L, 2);  // Use lua's concat operator (..)
+	lua_pushvalue(L, 2);  // Push current key
+	lua_pushnumber(L, 2527);
+
+        // cord.await(storm.os.invokeLater, 70*storm.os.MILLISECOND)
+	lua_getglobal(L, "cord");
```

Comment:
The call stack is lost when we yield, so we're going to have to use nc_set_continuation... which is going to require quite a bit of restructuring. We'll have to make the next key in the table an upvalue of the function that we invoke next, meaning that we'll probably have to change the for loop to something else.

I'm really curious how our test managed to work despite this error; it seems like it should have printed an error after the notification completed. I don't have the firestorm on me so I can't make any bugfixes at the moment. We'll probably have to wait until tomorrow to fix this...